### PR TITLE
Add: go safely

### PR DESCRIPTION
**Pull request by AlexStocks**
**What this PR does**:

run goroutime func safely

### CODE REVIEW

File: runtime/goroutine.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			if wg != nil {
+				wg.Done()
+			}
+		}()
+		handler()
+	}()
+}
+
+func GoUnterminal(wg *sync.WaitGroup, handle func()) {
+	GoSafely(wg,
```
Comment (flycash) [inline runtime/goroutine.go]: What will happen if `handler` always panic? I think the execution will be endless. handler panic -> final function -> handler panic -> final function ...

File: runtime/goroutine.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import (
+	"fmt"
+	"os"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ignoreRecover = false
```
Comment (flycash) [inline runtime/goroutine.go]: It seems that users cannot set the `ignoreRecover` to be true. So, why don't you remove this parameter? Or put this as `GoSafely` function's parameter?

File: runtime/goroutine_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		},
+		func(r interface{}) {
+			times++
+		},
+	)
+	time.Sleep(1e9)
+	assert.True(t, times == 3)
+}
+
+func TestGoUnterminal(t *testing.T) {
```
Comment (flycash) [inline runtime/goroutine_test.go]: Could you add one more test for `GoUnterminal` which `handler` always panic?

File: runtime/goroutine.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+// GoSafely wraps a `go func()` with recover()
+func GoSafely(wg *sync.WaitGroup, handler func(), finalFunc func(r interface{})) {
+	if wg != nil {
+		wg.Add(1)
+	}
+	go func() {
```
Comment (wongoo) [inline runtime/goroutine.go]: async run handler?

File: runtime/goroutine_test.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	GoUnterminal(
+		&wg,
+		func() {
+			if atomic.AddUint64(&times, 1) == 2 {
+				panic("hello")
+			}
+		},
+	)
+	wg.Wait()
+	assert.True(t, times == 3)
```
Comment (wongoo) [inline runtime/goroutine_test.go]: atomic load times

File: runtime/goroutine.go (hunk 1 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: yes, as its name said, this goroutine will run endless. This func is used for which goroutine wanna long live as its process.

File: runtime/goroutine.go (hunk 2 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: good idea.

File: runtime/goroutine.go (hunk 4 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: I do not know what u mean.

File: runtime/goroutine_test.go (hunk 5 above)
Comment (AlexStocks) [reply runtime/goroutine_test.go]: okay

**Comment by AlexStocks** [comment]
> I can't get what situation need `GoUnterminal`, while I think a retry mechanism with time limit would be better.

This func is used for which goroutine wanna long live as its process.

File: runtime/goroutine_test.go (hunk 3 above)
Comment (AlexStocks) [reply runtime/goroutine_test.go]: This test is enough.

File: runtime/goroutine.go (hunk 4 above)
Comment (wongoo) [reply runtime/goroutine.go]: `handler()` 默认异步执行是不是不友好？ 可以GoSafely同步执行 handler(), 如果有要异步执行的，用户自己`go GoSafely()`调用

File: runtime/goroutine.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		defer func() {
+			if r := recover(); r != nil {
+				if !ignoreRecover {
+					fmt.Fprintf(os.Stderr, "%s goroutine panic: %v\n%s\n",
+						time.Now(), r, string(debug.Stack()))
+				}
+				if finalFunc != nil {
```
Comment (wongoo) [inline runtime/goroutine.go]: `finalFunc` 是只有出现panic才调用吗？

File: runtime/goroutine.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+		handler()
+	}()
+}
+
+// GoUnterminated is used for which goroutine wanna long live as its process.
+func GoUnterminated(wg *sync.WaitGroup, ignoreRecover bool, handle func()) {
+	GoSafely(wg,
+		ignoreRecover,
+		func() {
+			handle()
```
Comment (wongoo) [inline runtime/goroutine.go]: handle本身就是func(), 不用包一层吧

File: runtime/goroutine.go (hunk 4 above)
Comment (wongoo) [reply runtime/goroutine.go]: 或者把方法名改为 Safely 即可？

File: runtime/goroutine.go (hunk 6 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: yeah.

File: runtime/goroutine.go (hunk 4 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: goSafely 的目的就是取代 go 调用，不让用户直接接触 go 这个原语。或者说，凡是dubbogo中出现go的地方，这个pr合并后，都会改成gosafely调用。

File: runtime/goroutine.go (hunk 7 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: 那你用测试用例，直接把handle放这里，看是否有语法错误提示。

File: runtime/goroutine.go (hunk 4 above)
Comment (wongoo) [reply runtime/goroutine.go]: 了解

File: runtime/goroutine.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import (
+	"fmt"
+	"os"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+// GoSafely wraps a `go func()` with recover()
+func GoSafely(wg *sync.WaitGroup, ignoreRecover bool, handler func(), finalFunc func(r interface{})) {
```
Comment (Patrick0308) [inline runtime/goroutine.go]: May be we need a time interval parameter? You can see this [JitterUntil Function](https://github.com/kubernetes/kubernetes/blob/86fcb30e5391a5494c8d83586f0911772a6949aa/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L132)

Comment (AlexStocks) [reply runtime/goroutine.go]: good idea.

**Pull request merged**